sql/parser/expr.go

```diff
@@ -241,72 +241,156 @@ func (NullVal) String() string {
 
 // QualifiedName is a base name and an optional indirection expression.
 type QualifiedName struct {
-	Base     Name
-	Indirect Indirection
+	Base       Name
+	Indirect   Indirection
+	tableName  bool
+	columnName bool
 }
 
 // StarExpr is a convenience variable that represents an unqualified "*".
 var StarExpr = &QualifiedName{Indirect: Indirection{unqualifiedStar}}
 
+// NormalizeTableName normalizes the qualified name to contain a database name
+// as prefix, returning an error if unable to do so or if the name is not a
+// valid table name (e.g. it contains an array indirection).
+func (n *QualifiedName) NormalizeTableName(database string) error {
+	if n == nil || n.Base == "" {
+		return fmt.Errorf("empty table name: %s", n)
+	}
+	// The qualified name should have the form "table", "database.table",
+	// "table@index" or "database.table@index".
```

Comment:
Yeah, the code changed and the comment no longer applies to the following if-statement. Let me rework the comments here and in `NormalizeColumnName`.